### PR TITLE
POST endpoint to rate place a11y

### PR DESCRIPTION
**Pull request by mutaphysis**
💥 Based on #21, ensure to merge/rebase in order

Created a new POST endpoint for rating the a11y of places.

This API behaves like the wheelmap classic API and abstracts a11yjson properties to simple mode. Allows rating wheelchair and toilet a11y, ranging from `yes`, `partial`, `no`, `unknown`. 

Places can only be rated when

 1. the source allows rating:
    - it has the `isA11yRatingAllowed` flag set to true
    - it is the default place adding source of the app matching the apptoken

 2. the source can be accessed
    - it is freely accessible, by setting `isFreelyAccessible` to true
    - the organization of the app of the appToken used is allowed access in `accessRestrictedTo`

The valid modes are `wheelchair` and `toilet`.
The valid ratings are  `yes`, `partial`, `no`, `unknown`. 

## Example requests

**Rate a places toilet a11y**

`curl -X POST 'http://localhost:4000/place-infos/rate?id=...&mode=toilet&rating=yes&appToken=...' -H 'Accept: application/json'`

**Response**
```
{"success":true}
```

**Rate a places wheelchair a11y**

`curl -X POST 'http://localhost:4000/place-infos/rate?id=..&mode=toilet&rating=unknown&appToken=...' -H 'Accept: application/json'`

**Response**
```
{"error":{"reason":"Place with id ... not found."}}
```

### CODE REVIEW

File: imports/server/json-api/PlaceInfos/report.ts [hunk 1]
```diff
@@ -38,6 +38,7 @@ WebApp.connectHandlers.use('/place-infos/report', (req: ParsedIncomingMessage, r
   const isAuthorized = isRequestAuthorized(req);
   if (!isAuthorized) {
     respondWithError(res, 401, 'Invalid token.');
+    return;
```
Comment (opyh) [inline imports/server/json-api/PlaceInfos/report.ts]: 🤭

File: imports/server/json-api/PlaceInfos/rate.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+  if (!place) {
+    throw new Meteor.Error(404, `Place with id ${placeId} not found.`);
+  }
+
+  const source = Sources.findOne(place.properties.sourceId);
+  if (!source) {
+    throw new Meteor.Error(404, `No source for place found.`);
+  }
+
+  const canAccessSource = source.organizationId === app.organizationId || source.isFreelyAccessible || source.accessRestrictedTo.indexOf(app.organizationId);
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/server/json-api/PlaceInfos/rate.ts]: Ups, this is very wrong, good catch

File: imports/server/collections/collections/PlaceInfos/createPlaceInfoOverApi.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+export function isInsertableViaAPI({ doc, appId, userId }: { doc: IPlaceInfo, appId?: string, userId?: string }) {
+  const app = Apps.findOne(appId);
+
+  if (!app) {
+    throw new Meteor.Error(404, `No app found for given appId ${appId}`);
+  }
+
+  const sourceId = app.defaultSourceIdForAddedPlaces;
+  if (!sourceId) {
+    throw new Meteor.Error(401, `No defaultSourceIdForAddedPlaces defined on app ${app.name}`);
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/server/collections/collections/PlaceInfos/createPlaceInfoOverApi.ts]: Done

File: imports/server/json-api/PlaceInfos/rate.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+  const query = parse(req.url, true).query;
+
+  const { app, appToken } = getAppAndUserFromRequest(req);
+  if (!app) {
+    throw new Meteor.Error(404, `No app found for given appToken ${appToken}`);
+  }
+
+  const sourceId = app.defaultSourceIdForAddedPlaces;
+  if (!sourceId) {
+    throw new Meteor.Error(401, `No defaultSourceIdForAddedPlaces defined on app ${app.name}`);
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/server/json-api/PlaceInfos/rate.ts]: Done

File: imports/server/json-api/PlaceInfos/rate.ts [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+  }
+
+  const canAccessSource = source.organizationId === app.organizationId || source.isFreelyAccessible || source.accessRestrictedTo.indexOf(app.organizationId); 
+  if (!canAccessSource) {
+    throw new Meteor.Error(401, `This appToken has no access to this places source. Please check with the owner of the source ${source.shortName}.`);
+  }
+
+  const allowRatingPlaces = place.properties.sourceId === sourceId || source.isA11yRatingAllowed;
+  if (!allowRatingPlaces) {
+    throw new Meteor.Error(401, `The place with the id ${placeId} cannot be rated via API as ${source.shortName} is neither modifyable or the default add place source.`);
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/server/json-api/PlaceInfos/rate.ts]: Done

File: imports/server/json-api/PlaceInfos/rate.ts [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      break;
+    }
+    case 'toilet': {
+      if (rating === 'unknown') {
+        updateData = hasRestrooms ? { 
+          $unset: { 
+            'properties.accessibility.restrooms.$[].isAccessibleWithWheelchair': 1,
+          } 
+        } : null;
+      } else if (rating === 'no' || rating === 'partial') {
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/server/json-api/PlaceInfos/rate.ts]: Added

File: imports/server/json-api/PlaceInfos/rate.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mutaphysis) [reply imports/server/json-api/PlaceInfos/rate.ts]: Fixed

**Pull request merged**